dbms/src/Flash/Mpp/GRPCReceiveQueue.h

```diff
@@ -0,0 +1,235 @@
+// receiving data are limited and the available grpc threads will be less and less when more
+// and more push operation are blocked. In order to avoid the above case, we create this
+// non-blocking queue so that push operation will not be blocked when the queue is full.
+template <typename T>
+class GRPCReceiveQueue
+{
+public:
+    GRPCReceiveQueue(size_t queue_size, grpc_call * call, const LoggerPtr & log_)
+        : recv_queue(queue_size)
+        , log(log_)
+        , kick_recv_tag(log)
+    {
+        RUNTIME_ASSERT(call != nullptr, log, "call is null");
+        // If a call to `grpc_call_start_batch` with an empty batch returns
+        // `GRPC_CALL_OK`, the tag is pushed into the completion queue immediately.
+        // This behavior is well-defined. See https://github.com/grpc/grpc/issues/16357.
+        kick_func = [call](void * t) {
+            return grpc_call_start_batch(call, nullptr, 0, t, nullptr);
+        };
+    }
+
+    // For gtest usage.
+    GRPCReceiveQueue(size_t queue_size, GRPCKickFunc func)
+        : recv_queue(queue_size)
+        , log(Logger::get())
+        , kick_func(func)
+        , kick_recv_tag(log)
+    {}
+
+    const String & getCancelReason() const
+    {
+        return recv_queue.getCancelReason();
+    }
+
+    // Cancel the send queue, and set the cancel reason
+    bool cancelWith(const String & reason)
+    {
+        auto ret = recv_queue.cancelWith(reason);
+        if (ret)
+            handleTheRemainingTags();
+        return ret;
+    }
+
+    bool finish()
+    {
+        auto ret = recv_queue.finish();
+        if (ret)
+            handleTheRemainingTags();
+        return ret;
+    }
+
+    // Pop data from queue and kick the CompletionQueue to re-push the data
+    // if there is data failing to be pushed before because the queue was full.
+    template <typename U>
+    bool pop(U && data)
+    {
+        auto ret = recv_queue.pop(std::forward<U>(data)) == MPMCQueueResult::OK;
+        if (ret)
+            kickCompletionQueue();
+        return ret;
+    }
+
+    // Push data into the queue.
+    //
+    // Return FULL if the queue is full and `new_tag` is saved.
+    // When the next pop/finish is called, the `new_tag` will be pushed
+    // into grpc completion queue.
+    template <typename U>
+    GRPCReceiveQueueRes push(U && data, void * new_tag)
+    {
+        auto res = recv_queue.tryPush(data);
+        switch (res)
+        {
+        case MPMCQueueResult::OK:
+            return GRPCReceiveQueueRes::OK;
+        case MPMCQueueResult::FINISHED:
+            return GRPCReceiveQueueRes::FINISHED;
+        case MPMCQueueResult::CANCELLED:
+            return GRPCReceiveQueueRes::CANCELLED;
+        case MPMCQueueResult::FULL:
+            // Handle this case later.
+            break;
+        default:
+            RUNTIME_ASSERT(false, log, "Result {} is invalid", static_cast<Int32>(res));
+        }
+
+        std::lock_guard lock(mu);
```

Comment:
I prefer to refine the code like 
```
MPMCQueueResult res;
for (size_t i = 0; i < 2; ++i)
{
    res = recv_queue.tryPush(data);
    switch (res)
    {
    case MPMCQueueResult::OK:
        return GRPCReceiveQueueRes::OK;
    case MPMCQueueResult::FINISHED:
        return GRPCReceiveQueueRes::FINISHED;
    case MPMCQueueResult::CANCELLED:
        return GRPCReceiveQueueRes::CANCELLED;
    case MPMCQueueResult::FULL:
        break;
    default:
        RUNTIME_ASSERT(false, log, "Result {} is invalid", static_cast<Int32>(res));
    }
}
assert(res == MPMCQueueResult::FULL);
{
    std::lock_guard lock(mu);
    tags.push_back(new_tag);
}
```

Reply:
And it seems that lock_guard only used for tags

Reply:
> And it seems that lock_guard only used for tags

The second tryPush must be contained in the lock.

If we don't contain the second tryPush in the lock, the following bug will happen:
Thread1: tryPush fail at first.
Thread1: tryPush fail at second.
Thread2: pop successfully and find that it's needless to kick CompletionQueue.
Thread1: lock and set tags. However, the pop of the Thread2 is the last pop operation, no more pop operation will be executed. So the tag will not be kicked into completion queue forever.

Reply:
ok, refine the code like
```
MPMCQueueResult res = recv_queue.tryPush(data);
if (res == MPMCQueueResult::FULL)
{
    // add some comment about why add lock where.
    std::lock_guard lock(mu);
	res = recv_queue.tryPush(data);
    if (res == MPMCQueueResult::FULL)
	    tags.push_back(new_tag);
}
switch (res)
{
case MPMCQueueResult::OK:
    return GRPCReceiveQueueRes::OK;
case MPMCQueueResult::FINISHED:
    return GRPCReceiveQueueRes::FINISHED;
case MPMCQueueResult::CANCELLED:
    return GRPCReceiveQueueRes::CANCELLED;
case MPMCQueueResult::FULL:
    return GRPCReceiveQueueRes::FULL;
default:
    RUNTIME_ASSERT(false, log, "Result {} is invalid", magic_enum::enum_name(res);
}
```
we can use magic_enum::enum_name to print enum, ref https://github.com/pingcap/tiflash/issues/5758

Reply:
> ok, refine the code like
> 
> ```
> MPMCQueueResult res = recv_queue.tryPush(data);
> if (res == MPMCQueueResult::FULL)
> {
>     // add some comment about why add lock where.
>     std::lock_guard lock(mu);
>     res = recv_queue.tryPush(data);
>     if (res == MPMCQueueResult::FULL)
> 	tags.push_back(new_tag);
> }
> switch (res)
> {
> case MPMCQueueResult::OK:
>     return GRPCReceiveQueueRes::OK;
> case MPMCQueueResult::FINISHED:
>     return GRPCReceiveQueueRes::FINISHED;
> case MPMCQueueResult::CANCELLED:
>     return GRPCReceiveQueueRes::CANCELLED;
> case MPMCQueueResult::FULL:
>     return GRPCReceiveQueueRes::FULL;
> default:
>     RUNTIME_ASSERT(false, log, "Result {} is invalid", magic_enum::enum_name(res);
> }
> ```
> 
> we can use magic_enum::enum_name to print enum, ref #5758

okk